// hexlet/dist/games/brain-balance.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _hexletPairs = require('hexlet-pairs');

var _ = require('..');

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var rule = 'Balance the given number.';

var valueRangeMin = 10;
var valueRangeMax = 1000;

var getMaxOfArray = function getMaxOfArray(numArray) {
  return Math.max.apply(null, numArray);
};

var getMinOfArray = function getMinOfArray(numArray) {
  return Math.min.apply(null, numArray);
};

var normalizeToString = function normalizeToString(arr) {
  var ascendingArr = arr.sort();
  return ascendingArr.join('');
};

var balanceArr = function balanceArr(arr) {
  var balancedArr = arr;

  for (var i = 0; i < balancedArr.length; i += 1) {
    for (var j = 0; j < balancedArr.length; j += 1) {
      if (balancedArr[i] - balancedArr[j] > 1) {
        balancedArr[i] -= 1;
        balancedArr[j] += 1;
      }
    }
  }
  return balancedArr;
};

var getBalancedAsString = function getBalancedAsString(n) {
  var str = String(n);
  var strArray = str.split('');
  var numArray = strArray.map(function (num) {
    return Number(num);
  });
  var balancedArray = balanceArr(numArray);
  while (getMaxOfArray(balancedArray) - getMinOfArray(balancedArray) > 1) {
    balancedArray = balanceArr(balancedArray);
  }
  return normalizeToString(balancedArray);
};

var askBalance = function askBalance() {
  var num = (0, _utils2.default)(valueRangeMin, valueRangeMax);
  var question = '' + num;
  var answer = getBalancedAsString(num);
  var game = (0, _hexletPairs.cons)(question, answer);
  return game;
};

var startBalanceGame = function startBalanceGame() {
  return (0, _.makeGame)(askBalance, rule);
};

exports.default = startBalanceGame;
